zug/util.hpp

```diff
@@ -112,6 +114,38 @@ struct composed
     }
 };
 
+template <typename T>
+constexpr bool is_composed_v = false;
+
+template <typename... Fns>
+constexpr bool is_composed_v<composed<Fns...>> = true;
+
+template <template <typename...> class Composed, typename... Fns>
+auto to_function_tuple(Composed<Fns...>&& c, meta::try_t)
+    -> std::enable_if_t<is_composed_v<std::decay_t<Composed<Fns...>>>,
+                        decltype(std::forward<Composed<Fns...>>(c).fns)>
+{
+    return std::forward<Composed<Fns...>>(c).fns;
+}
+
+template <typename Fn>
+auto to_function_tuple(Fn&& fn, meta::catch_t)
+{
+    return std::make_tuple(std::forward<Fn>(fn));
+}
+
+template <typename T>
+auto to_function_tuple(T&& t)
+{
+    return to_function_tuple(std::forward<T>(t), meta::try_t{});
+}
+
+template <template <typename...> class Tuple, typename... Fns>
+auto make_composed(Tuple<Fns...>&& fns)
```

Comment:
This is too generic, use `std::tuple` instead of parametrizing `Tuple<>`.

Reply:
> This is too generic, use `std::tuple` instead of parametrizing `Tuple<>`.

The idea here is to infer the value type of `Tuple` so it can be perfectly forwarded. `std::tuple` wouldn't support that - we'd have to `std::move` the parameter.

You mentioned in another comment that you weren't sure whether this does really perform perfect forwarding but I'm confident it does as it needs to infer the full type of `Tuple` (which could be `std::tuple&`, `std::tuple&&`, `const std::tuple&`). Currently the tests are only checking with rvalues, but I can add a couple of tests for lvalues to ensure this.

Reply:
Thanks for the clarification!

Reply:
I notice now it's only being called with an rvalue reference from `comp`. I personally prefer to keep the forwarding but I can remove it if you prefer.

Reply:
Hm, OK, forget all of this. I've been playing around more with compiler explorer and you're absolutely right - it isn't interpreted as a forwarding reference...